### ISSUE TITLE
IoU loss calculation

### ISSUE THREAD

**Issue by NiranjanRavi1993** [ultralytics/yolov5#6964]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi,
     In train.py, computeLoss function utilizes  bbox_iou  function in utils/metrics.py. There are different types such as IoU, GIoU, and DIoU loss metrics.  Once the training is completed, when we run the val.py, it uses the box_iou function (IoU loss metric) to provide results. 

My question is if we use the GIoU loss metric during training, then for evaluation,  box_iou in val.py should also be modified right? 
Or 
Will it be the same IoU threshold (box_iou) in val.py? 
Please correct me if I am wrong. 

Thanks and Regards
Niranjan Ravi.

**Comment by glenn-jocher** [comment]
@NiranjanRavi1993 nothing should be modified. Metrics are computed correctly the mAP@0.5:0.95 specifications.

**Comment by NiranjanRavi1993** [comment]
@glenn-jocher  Thank you for your reply. I am training YOLOv5 model using Pascal VOC dataset. During evaluation, I am able to observe results in mAP@.5 mAP@.5:.95 .  

What does **iou_thres** parameter in hyper-parameter signify? Default is 0.6. Does it mean mAP is calculated for 60% overlap? 

I would also like to see results in at various mAP such as 60%, 65%, 70% to understand the model performance. To achieve this, do I have to change the **iou_thres** everytime and re-run the model ? 

Please correct me if I am wrong. 

Thank you.

**Comment by glenn-jocher** [comment]
@NiranjanRavi1993 iou threshold hyperparameter is used for label-anchor matching.

You can view individual mAP threshold metrics in `ap`:
https://github.com/ultralytics/yolov5/blob/3f634d43c8ecea14aa9037e2fd28ded0433d491d/val.py#L253

**Comment by NiranjanRavi1993** [comment]
Thank you @glenn-jocher . I realized I did not give a path for save_dir before.  Below is the format, I am trying to achieve 

![image](https://user-images.githubusercontent.com/42325377/158862903-e4cc5133-6848-4990-bdd3-2790fae96ea8.png)

AP values at various thresholds.  COCO dataset gives AP for different objects automatically when we evaluate. For pascal, do we have to add any special conditions to observe the results similar to the above picture?

**Comment by glenn-jocher** [comment]
@NiranjanRavi1993 these are all in `ap` as I already mentioned above.

**Comment by NiranjanRavi1993** [comment]
@glenn-jocher  Thanks for the input. 

 If my understanding is right, mAP.5--> average precisions of all classes in the dataset with 50% threshold limit right? 

The things I experimented so far. 
1. Trained the YOLOV5n network on the PASCAL VOC dataset. 
2. Ran val.py and obtained mAP.5 as .8 and mAP.5:.95 at 54.8 (iou_t = .4)
3. Ran val.py again and obtained mAP.5 as .8 and mAP.5:.95 at 54.8 (iou_t = .6)

Question 1: Steps 2 and 3 are giving the same results? Does it mean iou_t is not taking any effect or how do I change the threshold ? (Quite not sure why this occurs)

Question2:  I looked into ap_per_class and gave the directory path to save the results. But couldn't see any files. My goal is find mAP at 60%, mAP at 65% until mAP 95% separately to observe the results. Can I achieve it with the val.py code? 

Thanks

**Comment by NiranjanRavi1993** [comment]
@glenn-jocher  Update: In this line at val.py, 
ap50, ap = ap[:, 0], ap.mean(1)  # AP@0.5, AP@0.5:0.95

If I choose ap[:, 1], it mean its **ap** calculated at **60% threshold**? It would be helpful to know if this is correct. 

Thank you.

**Comment by glenn-jocher** [comment]
@NiranjanRavi1993 

> Thank you @glenn-jocher . I realized I did not give a path for save_dir before. Below is the format, I am trying to achieve
> 
> ![image](https://user-images.githubusercontent.com/42325377/158862903-e4cc5133-6848-4990-bdd3-2790fae96ea8.png)
> 
> AP values at various thresholds. COCO dataset gives AP for different objects automatically when we evaluate. For pascal, do we have to add any special conditions to observe the results similar to the above picture?

The `ap` variable I indicated contains all the data you need for the above metrics and I already mentioned (twice) above.
